File: DAA/DAA python/DAA Pract2.py
from collections import Counter

class Node:
    def __init__(self, freq, symbol):
        self.left = None
        self.right = None
        self.freq = freq
        self.symbol = symbol

def huffman_encoding(data):
    freq = Counter(data)
    priority_queue = [Node(f, s) for s, f in freq.items()]
    while len(priority_queue) > 1:
        priority_queue = sorted(priority_queue, key=lambda x: x.freq)
        left = priority_queue.pop(0)
        right = priority_queue.pop(0)
        new_node = Node(left.freq + right.freq, None)
        new_node.left = left
        new_node.right = right
        priority_queue.append(new_node)
    
    root = priority_queue[0]
    codes = {}
    def generate_codes(node, code=''):
        if node:
            if node.symbol:
                codes[node.symbol] = code
            generate_codes(node.left, code + '0')
            generate_codes(node.right, code + '1')
    
    generate_codes(root)
    
    encoded_data = ''.join([codes[s] for s in data])
    
    return encoded_data, codes

def huffman_decoding(encoded_data, codes):
    reverse_codes = {v: k for k, v in codes.items()}
    decoded_data = ''
    code = ''
    for bit in encoded_data:
        code += bit
        if code in reverse_codes:
            decoded_data += reverse_codes[code]
            code = ''
    
    return decoded_data

# Example usage:
data = str (input("Enter the string: "))
#data = "huffman_encoding_example"
encoded_data, codes = huffman_encoding(data)
print("Encoded data:", encoded_data)
decoded_data = huffman_decoding(encoded_data, codes)
print("Decoded data:", decoded_data)

File: DAA/DAA cpp/DAA Pract2.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
using namespace std;

// Huffman tree node structure
struct HuffmanNode {
    char data;
    unsigned frequency;
    HuffmanNode *left, *right;

    HuffmanNode(char data, unsigned frequency) : data(data), frequency(frequency), left(nullptr), right(nullptr) {}
};

// Comparison function for the priority queue
struct CompareNodes {
    bool operator()(HuffmanNode* lhs, HuffmanNode* rhs) const {
        return lhs->frequency > rhs->frequency;
    }
};

// Function to generate Huffman codes for each character
void generateCodes(HuffmanNode* root, string code, unordered_map<char, string>& codes) {
    if (!root)
        return;

    if (root->data != '\0') {
        codes[root->data] = code;
    }

    generateCodes(root->left, code + "0", codes);
    generateCodes(root->right, code + "1", codes);
}

// Function to construct the Huffman tree and generate codes
void huffmanEncoding(const string& text) {
    unordered_map<char, unsigned> frequency;
    for (char ch : text) {
        frequency[ch]++;
    }

    priority_queue<HuffmanNode*, vector<HuffmanNode*>, CompareNodes> pq;

    for (const auto& pair : frequency) {
        pq.push(new HuffmanNode(pair.first, pair.second));
    }

    while (pq.size() != 1) {
        HuffmanNode *left = pq.top();
        pq.pop();

        HuffmanNode *right = pq.top();
        pq.pop();

        unsigned sumFreq = left->frequency + right->frequency;
        HuffmanNode* newNode = new HuffmanNode('\0', sumFreq);
        newNode->left = left;
        newNode->right = right;

        pq.push(newNode);
    }

    HuffmanNode* root = pq.top();

    unordered_map<char, string> codes;
    generateCodes(root, "", codes);

    cout << "Huffman Codes:" << endl;
    for (const auto& pair : codes) {
        cout << pair.first << " : " << pair.second << endl;
    }
}

int main() {
    string input;
    cout << "Enter a string: ";
    getline(cin, input);

    huffmanEncoding(input);

    return 0;
}
